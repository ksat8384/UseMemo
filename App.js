import React, { useState, useMemo } from "react";
import {
  SafeAreaView,
  StyleSheet,
  Text,
  TextInput,
  Button,
} from "react-native";
import List from "./src/components/List";

const users = [
  { id: "a", name: "Batman" },
  { id: "b", name: "Superman" },
  { id: "c", name: "Ironman" },
  { id: "d", name: "Hulk" },
  { id: "e", name: "BlackWidow" },
  { id: "f", name: "Thor" },
  { id: "g", name: "Captain America" },
  { id: "h", name: "Hawkeye" },
  { id: "i", name: "Spiderman" },
  { id: "j", name: "Antman" },
];

const App = () => {
  // console.log('Rendering App... ');
  const [text, setText] = useState("");
  const [search, setSearch] = useState("");

  const handleText = (text) => {
    console.log(text);
    setText(text);
  };

  const handleSearch = () => {
    setSearch(text);
  };

  /*
  Without useMemo, filter function will run for every key stroke the user types into the 
  input field eventhough filteredUsers only change when the search state changes and not 
  when text state changes.

  Try it yourself: Typing something into the input field shouldn't trigger the logging,
  but executing the search with the button click will trigger it.
  */
  const filteredUsers = useMemo(
    () =>
      users.filter((user) => {
        console.log("Filter function is running...");
        return user.name.toLowerCase().includes(search.toLowerCase());
      }),
    [search]
  );

  return (
    <SafeAreaView style={styles.container}>
      <TextInput
        testID="input"
        style={styles.input}
        placeholder="Enter search term here..."
        value={text}
        onChangeText={handleText}
      />

      <Button testID="search" title="Search" onPress={handleSearch} />

      <List list={filteredUsers} />
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 20,
  },
  input: {
    backgroundColor: "white",
    borderBottomColor: "#000000",
    borderBottomWidth: 1,
    marginBottom: 10,
  },
});

export default App;
