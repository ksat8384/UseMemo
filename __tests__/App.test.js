import React from "react";
import App from "../App";
import { render, cleanup } from "@testing-library/react-native";
import renderer from "react-test-renderer";

describe("App Component", () => {
  afterEach(cleanup);

  it("renders correctly", () => {
    const component = renderer.create(<App />);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
