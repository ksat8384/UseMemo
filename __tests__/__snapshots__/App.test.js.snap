// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
      "padding": 20,
    }
  }
>
  <TextInput
    onChangeText={[Function]}
    placeholder="Enter search term here..."
    style={
      Object {
        "backgroundColor": "white",
        "borderBottomColor": "#000000",
        "borderBottomWidth": 1,
        "marginBottom": 10,
      }
    }
    value=""
  />
  <View
    accessibilityRole="button"
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Search
      </Text>
    </View>
  </View>
  <RCTScrollView
    data={
      Array [
        Object {
          "id": "a",
          "name": "Batman",
        },
        Object {
          "id": "b",
          "name": "Superman",
        },
        Object {
          "id": "c",
          "name": "Ironman",
        },
        Object {
          "id": "d",
          "name": "Hulk",
        },
        Object {
          "id": "e",
          "name": "BlackWidow",
        },
        Object {
          "id": "f",
          "name": "Thor",
        },
        Object {
          "id": "g",
          "name": "Captain America",
        },
        Object {
          "id": "h",
          "name": "Hawkeye",
        },
        Object {
          "id": "i",
          "name": "Spiderman",
        },
        Object {
          "id": "j",
          "name": "Antman",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    testID="flat-list"
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#00ffff",
              "marginVertical": 8,
              "padding": 20,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 32,
              }
            }
          >
            Batman
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#00ffff",
              "marginVertical": 8,
              "padding": 20,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 32,
              }
            }
          >
            Superman
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#00ffff",
              "marginVertical": 8,
              "padding": 20,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 32,
              }
            }
          >
            Ironman
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#00ffff",
              "marginVertical": 8,
              "padding": 20,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 32,
              }
            }
          >
            Hulk
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#00ffff",
              "marginVertical": 8,
              "padding": 20,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 32,
              }
            }
          >
            BlackWidow
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#00ffff",
              "marginVertical": 8,
              "padding": 20,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 32,
              }
            }
          >
            Thor
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#00ffff",
              "marginVertical": 8,
              "padding": 20,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 32,
              }
            }
          >
            Captain America
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#00ffff",
              "marginVertical": 8,
              "padding": 20,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 32,
              }
            }
          >
            Hawkeye
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#00ffff",
              "marginVertical": 8,
              "padding": 20,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 32,
              }
            }
          >
            Spiderman
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#00ffff",
              "marginVertical": 8,
              "padding": 20,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 32,
              }
            }
          >
            Antman
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
